src/main/java/com/mongodb/jdbc/oidc/RFC8252HttpServer.java

```diff
@@ -0,0 +1,256 @@
+            Context context = new Context();
+            context.setVariable(PRODUCT_DOCS_LINK_KEY, PRODUCT_DOCS_LINK);
+            context.setVariable(PRODUCT_DOCS_NAME_KEY, PRODUCT_DOCS_NAME);
+            String acceptedHtml = templateEngine.process("OIDCAcceptedTemplate", context);
+            sendResponse(exchange, acceptedHtml, HttpURLConnection.HTTP_OK);
+            exchange.close();
+        }
+    }
+
+    /**
+     * Parses the query parameters from the HTTP exchange.
+     *
+     * @param exchange the HTTP exchange
+     * @return a map containing the parsed query parameters
+     * @throws UnsupportedEncodingException if the encoding is not supported
+     */
+    private Map<String, String> parseQueryParams(HttpExchange exchange)
+            throws UnsupportedEncodingException {
+        Map<String, String> queryParams = new HashMap<>();
+        String query = exchange.getRequestURI().getQuery();
+        if (query != null) {
+            for (String param : query.split("&")) {
+                String[] keyValue = param.split("=");
+                if (keyValue.length > 1) {
+                    queryParams.put(keyValue[0], URLDecoder.decode(keyValue[1], "UTF-8"));
+                } else {
+                    queryParams.put(keyValue[0], "");
+                }
+            }
+        }
+        return queryParams;
+    }
+
+    /**
+     * Puts the OIDC response into the blocking queue.
+     *
+     * @param exchange the HTTP exchange
+     * @param oidcResponse the OIDC response to put into the queue
+     * @return true if the response was successfully put into the queue, false otherwise
+     * @throws IOException if an I/O error occurs while sending a response
+     */
+    private boolean putOidcResponse(HttpExchange exchange, OidcResponse oidcResponse)
+            throws IOException {
+        try {
+            oidcResponseQueue.put(oidcResponse);
+            return true;
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            sendResponse(exchange, "<html><body><h1>Internal Server Error</h1></body></html>", 500);
+            return false;
+        }
+    }
+
+    /**
+     * Sends an HTTP response with the specified content and status code.
+     *
+     * @param exchange the HTTP exchange
+     * @param response the response content
+     * @param statusCode the HTTP status code
+     * @throws IOException if an I/O error occurs while sending the response
+     */
+    private void sendResponse(HttpExchange exchange, String response, int statusCode)
+            throws IOException {
+        exchange.getResponseHeaders().set("Content-Type", "text/html; charset=utf-8");
+        exchange.sendResponseHeaders(statusCode, response.getBytes(StandardCharsets.UTF_8).length);
+        OutputStream os = exchange.getResponseBody();
+        os.write(response.getBytes(StandardCharsets.UTF_8));
+        os.close();
```

Comment:
Do you need to close the outputstream or should you only close the channel?
In any case, I think you need to have try/catch or finally blocks here or in callers to make sure that you always close the exchange even when a exception is raised.
Otherwise, we will be leaking connections I think because the exchange will never be closed and so the outputstream and connection will stay open.

Reply:
I see.  I added a try with resource for the output stream and added a `finally` for closing the exchange.